lib/xboxrt/libc_extensions/direct.c

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: MIT
+
+// SPDX-FileCopyrightText: 2023 ExoSkye
+// SPDX-FileCopyrightText: 2025 Ryan Wendland
+
+// Part of Microsoft CRT
+
+#include <assert.h>
+#include <errno.h>
+#include <hal/debug.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <windows.h>
+
+// Made referencing https://www.digitalmars.com/rtl/direct.html, retrieved 2023-06-10, copyrighted 1999-2018 by Digital Mars
+
+// Only call this function in the case there actually is an error
+static int win32_error_to_errno (DWORD winerror)
+{
+    switch (winerror) {
+        case ERROR_FILE_NOT_FOUND:
+        case ERROR_PATH_NOT_FOUND:
+        case ERROR_FILENAME_EXCED_RANGE:
+        case ERROR_BAD_PATHNAME:
+            return ENOENT;
+        case ERROR_ALREADY_EXISTS:
+        case ERROR_FILE_EXISTS:
+            return EEXIST;
+        case ERROR_NOT_LOCKED:
+        case ERROR_ACCESS_DENIED:
+            return EACCES;
+        case ERROR_INVALID_PARAMETER:
+        case ERROR_INVALID_FUNCTION:
+            return EINVAL;
+        case ERROR_NOT_ENOUGH_MEMORY:
+            return ENOMEM;
+        case ERROR_DIR_NOT_EMPTY:
+            return ENOTEMPTY;
+        case ERROR_CALL_NOT_IMPLEMENTED:
+            return ENOSYS;
+        default:
+            debugPrint("WARNING: Unknown win32 error code 0x%08lx, returning EINVAL", winerror);
```

Comment:
I'd prefer if this was an assert instead, depending on the application this debugPrint might not go anywhere